src/D2L.Security.OAuth2/Keys/KeyManagementService.cs

```diff
@@ -0,0 +1,207 @@
+			IPublicKeyDataProvider publicKeys,
+			IPrivateKeyDataProvider privateKeys,
+			IDateTimeProvider clock,
+			OAuth2Configuration config
+		) {
+			m_publicKeys = publicKeys;
+			m_privateKeys = privateKeys;
+			m_clock = clock;
+			m_config = config;
+		}
+
+		async Task<D2LSecurityToken> IPrivateKeyProvider.GetSigningCredentialsAsync() {
+			var current = Volatile.Read( ref m_current );
+
+			var now = m_clock.UtcNow;
+
+			if ( current == null || current.ValidTo <= now ) {
+				// Slow path: RefreshKeyAsync() wasn't called on boot and/or it
+				// isn't being called in a background job.
+				await RefreshKeyAsync( current, now )
+					.ConfigureAwait( false );
+
+				current = Volatile.Read( ref m_current );
+			}
+
+			return current.Ref();
+		}
+
+		async Task<TimeSpan> IKeyManagementService.RefreshKeyAsync() {
+			var current = Volatile.Read( ref m_current );
+
+			var now = m_clock.UtcNow;
+
+			await RefreshKeyAsync(
+				current,
+				now
+			).ConfigureAwait( false );
+
+			current = Volatile.Read( ref m_current );
+
+			var expectedNextRotation = current.ValidTo - m_config.KeyRotationBuffer;
+
+			if( now > expectedNextRotation ) {
+				// If that's in the past, use a short retry window.
+				return TimeSpan.FromSeconds( 10 );
+			} else {
+				// Otherwise use that but with a little buffer for key
+				// generation time/imprecisely scheduled cron jobs.
+				return expectedNextRotation.AddMinutes( 1 ) - now;
+			}
+		}
+
+		async Task IKeyManagementService.GenerateNewKeyIfNeededAsync() {
+			var now = m_clock.UtcNow;
+
+			var keys = await m_privateKeys.GetAllAsync(
+				validUntilAtLeast: now
+			).ConfigureAwait( false );
+
+			foreach( var key in keys ) {
+				if( now < key.ExpiresAt - m_config.KeyRotationBuffer ) {
+					// This key is fine, do nothing.
+					return;
+				}
+			}
+
+			var keyId = Guid.NewGuid();
+			var keyIdStr = keyId.ToString();
+
+			using var csp = new RSACryptoServiceProvider {
+				PersistKeyInCsp = false
+			};
+
+			var pub = csp.ExportParameters(
+				includePrivateParameters: false
+			);
+
+			var priv = csp.ExportCspBlob(
+				includePrivateParameters: true
+			);
+
+			var expiresAt = now + m_config.KeyLifetime;
+
+			await Task.WhenAll(
+				m_publicKeys.SaveAsync(
+					keyId,
+					new RsaJsonWebKey(
+						keyIdStr,
+						expiresAt,
+						pub
+					)
+				),
+				m_privateKeys.SaveAsync(
+					new PrivateKeyData(
+						id: keyIdStr,
+						kind: PrivateKeyData.KeyKinds.Rsa,
+						data: priv,
+						createdAt: now,
+						notBefore: now + m_config.KeyTimeUntilUse,
+						expiresAt: expiresAt
+					)
+				)
+			).ConfigureAwait( false );
+		}
+
+		private async Task RefreshKeyAsync(
+			D2LSecurityToken current,
+			DateTimeOffset now
+		) {
+			var keys = await m_privateKeys.GetAllAsync(
+				validUntilAtLeast: now
+			).ConfigureAwait( continueOnCapturedContext: false );
+
+			var best = ChooseKey( keys, now ).Ref();
+
+			var loser = Interlocked.CompareExchange(

```

Comment:
`CompareExchange` doesn't return the loser. It always returns the original value of `m_current`

Reply:
Good call -- good behaviour to check with tests.

Changed it to Exchange which changes the intended behaviour to always using the "last" key we pick under contention which is fine. The original thought with CompareExchange (modulo correct disposal) was to avoid changing our mind and signing tokens with a different key suddenly, but (1) if there were a different choice of key so suddenly that is ... worrying (2) we have more than one server, so different choices would be made anyway.

(Every time we use a private key we force people to fetch the corresponding public key.)